mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAO.java

```diff
@@ -48,58 +50,106 @@
+        incrementMessageCountStatement = updateMailboxStatement(session, incr(COUNT));
+        incrementUnseenCountStatement = updateMailboxStatement(session, incr(UNSEEN));
+        addToCounters = session.prepare(update(TABLE_NAME)
+            .with(incr(COUNT, bindMarker(COUNT)))
+            .and(incr(UNSEEN, bindMarker(UNSEEN)))
+            .where(eq(MAILBOX_ID, bindMarker(MAILBOX_ID))));
+        removeToCounters = session.prepare(update(TABLE_NAME)
+            .with(decr(COUNT, bindMarker(COUNT)))
+            .and(decr(UNSEEN, bindMarker(UNSEEN)))
+            .where(eq(MAILBOX_ID, bindMarker(MAILBOX_ID))));
+        decrementMessageCountStatement = updateMailboxStatement(session, decr(COUNT));
+        decrementUnseenCountStatement = updateMailboxStatement(session, decr(UNSEEN));
     }
 
     private PreparedStatement createReadStatement(Session session) {
         return session.prepare(
-            select(CassandraMailboxCountersTable.UNSEEN, CassandraMailboxCountersTable.COUNT)
-                .from(CassandraMailboxCountersTable.TABLE_NAME)
-                .where(eq(CassandraMailboxCountersTable.MAILBOX_ID, bindMarker(CassandraMailboxCountersTable.MAILBOX_ID))));
+            select(UNSEEN, COUNT)
+                .from(TABLE_NAME)
+                .where(eq(MAILBOX_ID, bindMarker(MAILBOX_ID))));
     }
 
     private PreparedStatement updateMailboxStatement(Session session, Assignment operation) {
         return session.prepare(
-            update(CassandraMailboxCountersTable.TABLE_NAME)
+            update(TABLE_NAME)
                 .with(operation)
-                .where(eq(CassandraMailboxCountersTable.MAILBOX_ID, bindMarker(CassandraMailboxCountersTable.MAILBOX_ID))));
+                .where(eq(MAILBOX_ID, bindMarker(MAILBOX_ID))));
     }
 
     public Mono<MailboxCounters> retrieveMailboxCounters(CassandraId mailboxId) {
         return cassandraAsyncExecutor.executeSingleRow(bindWithMailbox(mailboxId, readStatement))
             .map(row ->  MailboxCounters.builder()
                 .mailboxId(mailboxId)
-                .count(row.getLong(CassandraMailboxCountersTable.COUNT))
-                .unseen(row.getLong(CassandraMailboxCountersTable.UNSEEN))
+                .count(row.getLong(COUNT))
+                .unseen(row.getLong(UNSEEN))
                 .build());
     }
 
-    public Mono<Long> countMessagesInMailbox(Mailbox mailbox) throws MailboxException {
+    public Mono<Void> resetCounter(MailboxCounters counters) {
+        CassandraId mailboxId = (CassandraId) counters.getMailboxId();
+
+        return retrieveMailboxCounters(mailboxId)
+            .switchIfEmpty(Mono.just(emptyCounters(mailboxId)))
+            .flatMap(storedCounters -> {
+                if (storedCounters.equals(counters)) {
+                    return Mono.empty();
+                }
+                return remove(storedCounters)
+                    .then(add(counters));
```

Comment:
Is it not risky?

Reply:
It's impossible to do better with counters